File: src/trading/LivePositionReconciliation.cpp
#include "trading/LivePositionReconciliation.hpp"

namespace trade {
namespace trading {

std::string livePositionReconciliationStatus(const LivePositionReconciliation &state) {
  if (state.snapshot_present) {
    return "coinbase_confirmed";
  }
  if (state.pending_order_present) {
    return "pending_settlement";
  }
  if (state.managed_quantity_floor_present) {
    return "awaiting_snapshot_reconciliation";
  }
  return state.snapshot_loaded ? "unverified_missing_from_snapshot"
                               : "unverified_no_snapshot";
}

bool livePositionContributesToExposure(const LivePositionReconciliation &state) {
  return state.snapshot_present || state.pending_order_present ||
         state.managed_quantity_floor_present;
}

} // namespace trading
} // namespace trade

File: frontend/lib/livePositionReconciliation.ts
// Mirrors the `reconciliation_status` values produced by
// `livePositionReconciliationStatus()` in
// src/trading/LivePositionReconciliation.cpp. The backend already excludes
// the two `unverified_*` states from exposure/PnL math via
// `livePositionContributesToExposure()`; this module lets the frontend apply
// the same exclusion to displayed counts and gate destructive actions on
// positions Coinbase hasn't confirmed exist.
export type LivePositionReconciliationStatus =
  | 'coinbase_confirmed'
  | 'pending_settlement'
  | 'awaiting_snapshot_reconciliation'
  | 'unverified_missing_from_snapshot'
  | 'unverified_no_snapshot';

const UNVERIFIED_STATUSES = new Set<string>([
  'unverified_missing_from_snapshot',
  'unverified_no_snapshot',
]);

export function isUnverifiedReconciliationStatus(status?: string | null): boolean {
  return Boolean(status) && UNVERIFIED_STATUSES.has(status as string);
}

const RECONCILIATION_STATUS_LABELS: Record<string, string> = {
  pending_settlement: 'Pending settlement',
  awaiting_snapshot_reconciliation: 'Awaiting Coinbase snapshot',
  unverified_missing_from_snapshot: 'Reconciliation pending',
  unverified_no_snapshot: 'Reconciliation pending',
};

export function reconciliationStatusLabel(status?: string | null): string | null {
  if (!status) return null;
  return RECONCILIATION_STATUS_LABELS[status] ?? null;
}
